File: WordPressKit/WordPressKit/RemoteBlogOptionsHelper.m
#import "RemoteBlogOptionsHelper.h"
#import "NSMutableDictionary+Helpers.h"
#import <WordPressKit/WordPressKit-Swift.h>
@import NSObject_SafeExpectations;
@import WordPressShared;

@implementation RemoteBlogOptionsHelper

// Formats blog options retrieved from REST queries
+ (NSDictionary *)mapOptionsFromResponse:(NSDictionary *)response
{
    NSMutableDictionary *options = [NSMutableDictionary dictionary];
    options[@"home_url"] = response[@"URL"];
    if ([[response numberForKey:@"jetpack"] boolValue]) {
        options[@"jetpack_client_id"] = [response numberForKey:@"ID"];
    }
    if ( response[@"options"] ) {
        options[@"post_thumbnail"] = [response valueForKeyPath:@"options.featured_images_enabled"];
        NSArray *optionsDirectMapKeys = @[
                                          @"active_modules",
                                          @"admin_url",
                                          @"login_url",
                                          @"image_default_link_type",
                                          @"software_version",
                                          @"videopress_enabled",
                                          @"timezone",
                                          @"gmt_offset",
                                          @"allowed_file_types",
                                          @"frame_nonce",
                                          @"blog_public"
                                          ];

        for (NSString *key in optionsDirectMapKeys) {
            NSString *sourceKeyPath = [NSString stringWithFormat:@"options.%@", key];
            if ([response valueForKeyPath:sourceKeyPath] != nil) {
                options[key] = [response valueForKeyPath:sourceKeyPath];
            }
        }
    }
    NSMutableDictionary *valueOptions = [NSMutableDictionary dictionaryWithCapacity:options.count];
    [options enumerateKeysAndObjectsUsingBlock:^(id key, id obj, BOOL *stop) {
        valueOptions[key] = @{@"value": obj};
    }];

    return [NSDictionary dictionaryWithDictionary:valueOptions ];
}

+ (NSDictionary *)remoteOptionsForUpdatingBlogTitleAndTagline:(RemoteBlogSettings *)blogSettings
{
    NSMutableDictionary *options = [NSMutableDictionary dictionary];
    [options setValueIfNotNil:blogSettings.name forKey:@"blog_title"];
    [options setValueIfNotNil:blogSettings.tagline forKey:@"blog_tagline"];
    return options;
}

+ (RemoteBlogSettings *)remoteBlogSettingsFromXMLRPCDictionaryOptions:(NSDictionary *)options
{
    RemoteBlogSettings *remoteSettings = [RemoteBlogSettings new];
    remoteSettings.name = [[options stringForKeyPath:@"blog_title.value"] stringByDecodingXMLCharacters];
    remoteSettings.tagline = [[options stringForKeyPath:@"blog_tagline.value"] stringByDecodingXMLCharacters];
    if (options[@"blog_public"]) {
        remoteSettings.privacy = [options numberForKeyPath:@"blog_public.value"];
    }
    return remoteSettings;
}

@end

File: WordPressKit/Sources/WordPressKit/Models/RemoteBlogOptionsHelper.swift
import Foundation

@objcMembers public class RemoteBlogOptionsHelper: NSObject {

    public class func mapOptions(fromResponse response: NSDictionary) -> NSDictionary {
        let options = NSMutableDictionary()
        options["home_url"] = response["URL"]
        if response.number(forKey: "jetpack")?.boolValue == true {
            options["jetpack_client_id"] = response.number(forKey: "ID")
        }
        if response["options"] != nil {
            options["post_thumbnail"] = response.value(forKeyPath: "options.featured_images_enabled")

            let optionsDirectMapKeys = [
                "active_modules",
                "admin_url",
                "login_url",
                "unmapped_url",
                "image_default_link_type",
                "software_version",
                "videopress_enabled",
                "timezone",
                "gmt_offset",
                "allowed_file_types",
                "frame_nonce",
                "jetpack_version",
                "is_automated_transfer",
                "blog_public",
                "max_upload_size",
                "is_wpcom_atomic",
                "is_wpforteams_site",
                "show_on_front",
                "page_on_front",
                "page_for_posts",
                "blogging_prompts_settings",
                "jetpack_connection_active_plugins",
                "can_blaze"
            ]
            for key in optionsDirectMapKeys {
                if let value = response.value(forKeyPath: "options.\(key)") {
                    options[key] = value
                }
            }
        }
        let valueOptions = NSMutableDictionary(capacity: options.count)
        for (key, obj) in options {
            valueOptions[key] = [
                "value": obj
            ]
        }

        return NSDictionary(dictionary: valueOptions)
    }

    // Helper methods for converting between XMLRPC dictionaries and RemoteBlogSettings
    // Currently, we are only ever updating the blog title or tagline through XMLRPC
    // Brent - Jan 7, 2017
    public class func remoteOptionsForUpdatingBlogTitleAndTagline(_ blogSettings: RemoteBlogSettings) -> NSDictionary {
        let options = NSMutableDictionary()
        if let value = blogSettings.name {
            options["blog_title"] = value
        }
        if let value = blogSettings.tagline {
            options["blog_tagline"] = value
        }
        return options
    }

    public class func remoteBlogSettings(fromXMLRPCDictionaryOptions options: NSDictionary) -> RemoteBlogSettings {
        let remoteSettings = RemoteBlogSettings()
        remoteSettings.name = options.string(forKeyPath: "blog_title.value")?.stringByDecodingXMLCharacters()
        remoteSettings.tagline = options.string(forKeyPath: "blog_tagline.value")?.stringByDecodingXMLCharacters()
        if options["blog_public"] != nil {
            remoteSettings.privacy = options.number(forKeyPath: "blog_public.value")
        }
        return remoteSettings
    }

}
